### DISCUSSION TITLE
End to End security discussion

### DISCUSSION THREAD

**Discussion by 9to1url** [narrowlink/narrowlink#15] [General]
I double check what generate on Agent and Client side, the agent.yaml is just a JWT token from token-generator.yaml,
 I can't find a pair of public/private key been used for mitigate the MitM :

```
End-to-End Authentication: WireGuard uses pre-shared public keys for end-to-end authentication. 
Each party has a pair of public/private keys, and only devices with the correct public key can establish a connection. 
This ensures that a man-in-the-middle cannot insert or tamper with the communication without the correct keys.
```

Could you advise on this E2EE concern? thanks

**Comment by SajjadPourali** [comment]
The end-to-end encryption (E2EE) in Narrowlink differs from WireGuard. Narrowlink follows a distinct approach, as discussed [here](https://github.com/narrowlink/narrowlink/discussions/12#discussioncomment-6736259), and partially described in the [End-to-End Encryption](https://narrowlink.com/docs/extended-tutorial/end-to-end-encryption) guide (However can be improvemented).

Narrowlink utilizes a preshared secret derived from the users' defined passphrase,  which is not part of the token. Users are required to use a passphrase on both the client and agent sides. Subsequently, this passphrase is hashed using Sha3-256 to generate the encryption key.

**Comment by SajjadPourali** [comment]
I have added more clarification to the docs https://github.com/narrowlink/homepage/commit/86b22529d8577048b3c7ec41a107d2c4a514fd3b
